File: LeetCode/148.Sort_List/Mario/148.Sort_List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        // split two list
        ListNode right = getMid(head);
        ListNode left = head;
        ListNode tmp = right.next;
        right.next = null;
        right = tmp;

        left = sortList(left);
        right = sortList(right);

        return merge(left, right);
    }

    private ListNode getMid(ListNode head) {
        ListNode fast = head.next;
        ListNode slow = head;
        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }
        return slow;
    }

    private ListNode merge(ListNode left, ListNode right) {
        ListNode dummy = new ListNode();
        ListNode node = dummy;

        while (left != null && right != null) {
            if (left.val > right.val) {
                node.next = right;
                right = right.next;
            } else {
                node.next = left;
                left = left.next;
            }
            node = node.next;
        }

        if (left != null) {
            node.next = left;
        }

        if (right != null) {
            node.next = right;
        }

        return dummy.next;
    }

}

File: LeetCode/148.Sort_List/Mario/148.Sort_List.py
#!/usr/bin/env python3


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # divide and conquer
        if not head or not head.next:
            return head
        # break down into two linked lists
        dummy = ListNode(-1)
        dummy.next = head
        slow, fast = dummy, dummy
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        right = slow.next
        slow.next = None

        l = self.sortList(dummy.next)
        r = self.sortList(right)

        # merge
        return self.merge(l, r)

    def merge(self, l, r):
        if not l and not r:
            return None

        dummy = ListNode(-1)
        p = dummy
        while l and r:
            if l.val < r.val:
                p.next = l
                l = l.next
            else:
                p.next = r
                r = r.next
            p = p.next

        if l:
            p.next = l
        if r:
            p.next = r

        return dummy.next
